frontend/src/common/SortedCollection.ts

```diff
@@ -97,27 +97,44 @@ export namespace SortedCollection {
   export type StateRef = Opaque<number, 'SortedCollection.StateRef'>;
 }
 
+interface Focus {
+  start: number;
+  end: number;
+}
+
 export class SortedCollection<Item extends { id: number }> {
+  // Comparator function used to sort the collection
   private compare: Compare<Item>;
+  // Mapping item `id` to the `Item`, this uses array as a structure with
+  // the assumption that `id`s provided are increments from `0`, and that
+  // vacant `id`s will be re-used in the future.
   private map = Array<Maybe<Item>>();
+  // Actual sorted list of `Item`s.
   private list = Array<Item>();
+  // Internal tracker for changes, this number increments whenever the
+  // order of the **focused** elements in the collection changes
   private changeRef = 0;
+  // Marks the range of indicies that are focused for tracking.
+  // **Note:** `start` is inclusive, while `end` is exclusive (much like
+  // `Array.slice()`).
+  private focus: Focus = { start: 0, end: 0 };
 
   constructor(compare: Compare<Item>) {
```

Comment:
Will give this a closer read tomorrow, but just a small syntax sugar thing that I like in typescript is that you can drop the body of the constructor here and the member declaration by just having:

```javascript
constructor(private compare: Compare<Item>) {}
```
Read more here: https://www.typescriptlang.org/docs/handbook/classes.html#parameter-properties

That being said, having the comment above member declaration is nice and I think this is really just an opinion thing